include/deal.II/integrators/grad_div.h

```diff
@@ -0,0 +1,379 @@
+            {
+              double dv = 0.;
+              double du = 0.;
+              for (unsigned int d=0; d<dim; ++d)
+                {
+                  dv += fetest.shape_grad_component(i,k,d)[d];
+                  du += input[d][k][d];
+                }
+
+              result(i) += dx * du * dv;
+            }
+        }
+    }
+
+    /**
+     * The matrix for the weak boundary condition of Nitsche type for linear elasticity:
+     * @f[
+     * \int_F \Bigl(\gamma (u \cdot n)(v \cdot n)  - \nabla\cdot u
+     * v\cdot n - u \cdot n \nabla \cdot v \Bigr)\;ds.
+     * @f]
+     */
+    template <int dim>
+    inline void nitsche_matrix (
+      FullMatrix<double> &M,
+      const FEValuesBase<dim> &fe,
+      double penalty,
+      double factor = 1.)
+    {
+      const unsigned int n_dofs = fe.dofs_per_cell;
+
+      AssertDimension(fe.get_fe().n_components(), dim);
+      AssertDimension(M.m(), n_dofs);
+      AssertDimension(M.n(), n_dofs);
+
+      for (unsigned int k=0; k<fe.n_quadrature_points; ++k)
+        {
+          const double dx = factor * fe.JxW(k);
+          const Tensor<1,dim> n = fe.normal_vector(k);
+          for (unsigned int i=0; i<n_dofs; ++i)
+            for (unsigned int j=0; j<n_dofs; ++j)
+              {
+                double un = 0., vn = 0., divu = 0., divv = 0.;
+                for (unsigned int d=0; d<dim; ++d)
+                  {
+                    un += fe.shape_value_component(j,k,d) * n[d];
+                    vn += fe.shape_value_component(i,k,d) * n[d];
+                    divu += fe.shape_grad_component(j,k,d)[d];
+                    divv += fe.shape_grad_component(i,k,d)[d];
+                  }
+
+                M(i,j) += dx * 2. * penalty * un * vn;
+                M(i,j) -= dx*(divu*vn+divv*un);
+              }
+        }
+    }
+
+    /**
+     * Weak boundary condition for the Laplace operator by Nitsche, vector
+     * valued version, namely on the face <i>F</i> the vector
+     * @f[
+     * \int_F \Bigl(\gamma (\mathbf u \cdot \mathbf n- \mathbf g \cdot
+     * \mathbf n) (\mathbf v \cdot \mathbf n)
+     * - \nabla \cdot \mathbf u (\mathbf v \cdot \mathbf n)
+     * - (\mathbf u-\mathbf g) \cdot \mathbf n \nabla \cdot v\Bigr)\;ds.
+     * @f]
+     *
+     * Here, <i>u</i> is the finite element function whose values and gradient
+     * are given in the arguments <tt>input</tt> and <tt>Dinput</tt>,
+     * respectively. <i>g</i> is the inhomogeneous boundary value in the
+     * argument <tt>data</tt>. $\gamma$ is the usual penalty parameter.
+     *
+     * @author Guido Kanschat
+     * @date 2008, 2009, 2010
+     */
+    template <int dim>
+    void nitsche_residual (
+      Vector<double> &result,
+      const FEValuesBase<dim> &fe,
+      const VectorSlice<const std::vector<std::vector<double> > > &input,
+      const VectorSlice<const std::vector<std::vector<Tensor<1,dim> > > > &Dinput,
+      const VectorSlice<const std::vector<std::vector<double> > > &data,
```

Comment:
We should get rid of all of these uses of `VectorSlice` by using `ArrayView` instead. But that's for another day.